rfd/0001-testing-guidelines.md

```diff
@@ -0,0 +1,197 @@
+---
+authors: Andrew Lytvynov (andrew@gravitational.com)
+state: draft
+---
+
+# RFD 1 - Testing guidelines
+
+## What
+
+Guidelines for writing Go tests for teleport.
+
+## Why
+
+At the time of writing this, teleport tests use
+[gocheck](https://labix.org/gocheck). This framework has several downsides:
+- all tests are serial - slow execution in packages with many small tests
+- easy to mess up setup (forgetting to call `check.T`) or run tests multiple
+  times (calling `check.T` multiple times)
+- limited deep-diffing of values
+- largely unmaintained
+
+At the same time, there are some features we'd like to retain:
+- organization of tests into testsuites
+- shared setup/teardown code between tests
+- asserts for concise validation
+
+Go's builtin `testing` package covers most of our needs, and gets new features
+with every new Go release. We just need to complement it with:
+- asserts
+- flexible value diffing (got vs want)
+
+## Details
+
+Use the `testing` package as a base and complement it with several libraries.
+
+### Test organization
+
+Use a hierarchy of:
+- `_test.go` files
+- `func Test*` functions
+- [`t.Run`](https://golang.org/pkg/testing/#hdr-Subtests_and_Sub_benchmarks)
+  subtests (aka table-driven tests)
+
+Shared setup/teardown:
+- [`func TestMain`](https://golang.org/pkg/testing/#hdr-Main) for package level
+- a single `func Test*` with subtests for smaller test groups
+- `func setupX` called from multiple `func Test*` for larger test groups
+
+For performance testing use
+[benchmarks](https://golang.org/pkg/testing/#hdr-Benchmarks).
+
+### Parallelization
+
+Call [`t.Parallel()`](https://golang.org/pkg/testing/#T.Parallel) from all
+tests and subtests that don't need to run serially.
+
+### Asserts
+
+Use
+[testify/assert](https://pkg.go.dev/github.com/stretchr/testify/assert?tab=doc)
+or a plain `if` condition, whichever is easiest.
+
+Skim through the `testify/assert` docs, it has many convenient helpers.
+
+### Diffing
+
+For non-trivial comparisons (such as deeply-nested structs or protobufs), use
+[go-cmp](https://pkg.go.dev/github.com/google/go-cmp/cmp?tab=doc) with
+[cmpopts](https://pkg.go.dev/github.com/google/go-cmp@v0.5.1/cmp/cmpopts?tab=doc).
+
+It allows you to customize checking by ignoring types/fields, equating empty
+and nil slices, approximating float comparisons. See examples below.
+
+### Wishlist
+
+Some things we don't have a good solution for yet:
+- log output collection
+  - if something is logged by a function under test and that test fails, print
+    the log output
+  - it's currently hard to separate logging from a function under test and
+    unrelated logs by a background goroutine
+- performance regression tracking
+  - `go test` will report runtimes of individual tests/subtests
+  - but we don't have tooling to track those over time
```

Comment:
I wonder if we could outsource this to one of the cloud-based testing services like [Codecov](https://codecov.io/) or something similar - something that might be able to more easily track trends over time.

Reply:
We could do that for coverage, but I'm not aware of anything off-the-shelf for performance.
Assuming we don't decide to focus on code coverage (IMO it's a questionable metric), let's punt this for now.